### PR TITLE
#315 Manage the content displayed on the "Stakeholder Directory

### PR DESCRIPTION
**Pull request by fgkinus**
**What does this PR do?**

 - Add functionality fo adding new stakeholders
- add functionality for editing stakeholders
- add functionality for editing stakeholders

**description of Task to be completed?**
- add new stakeholders to the directory
- consume dropdown data from the database

**How should this be manually tested?**
```
git clone https://github.com/andela/mmdp_site
cd mmdp_site
git checkout ft-add-stakeholders
yarn run: server
yarn client:dev
```

- browse to `<localhost>:<port>/stakeholder-directory/add-basic-information`

**Any background context you want to provide?**

**What are the relevant Jira issues?**

[MMDP-315](https://viisaus.atlassian.net/browse/MMDP-315)

**Screenshots (if appropriate)**
- add basic information view
image
 -add beneficiary information
image

### CODE REVIEW

File: server/helpers/stakeholdersDirectory/stakeholderHelpers.js [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+    stakeholder = await Stakeholder.findOneAndUpdate(
+      { _id: basicInformation._id },
+      basicInformation,
+      options,
+    );
+  } else {
+    await Stakeholder.findOne({
+      organisationName: basicInformation.organisationName,
+    })
+      .session(options.session) // if a user already exists, raise an error
```
Comment (shemogumbe) [inline server/helpers/stakeholdersDirectory/stakeholderHelpers.js]: This ought to be removed now that we are not working with transactions

File: server/helpers/stakeholdersDirectory/stakeholderHelpers.js [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+        return Promise.all([
+          addBeneficiaries(stakeholder._id, beneficiaries, opts),
+          addStakeholderAddress(stakeholder._id, address, opts),
+          addPartnerships(stakeholder._id, partnerships, opts),
+        ]).then((values) => {
+          return [stakeholder, values];
+        });
+      })
+      .then(async (results) => {
+        console.log('ending transaction');
```
Comment (shemogumbe) [inline server/helpers/stakeholdersDirectory/stakeholderHelpers.js]: Remove console log statements from the source code

File: server/helpers/stakeholdersDirectory/stakeholderHelpers.js [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+export const handleCreateUpdateTransaction = async (
+  basicInformation,
+  address = [],
+  partnerships = [],
+  beneficiaries = [],
+) => {
+  // this is a function to handle a single stakeholderTransaction.
+  // start a session and a transaction first
+  const session = await mongoose.startSession();
+  session.startTransaction();
```
Comment (shemogumbe) [inline server/helpers/stakeholdersDirectory/stakeholderHelpers.js]: Do not se transactions here, we can use independent functions with await and appropriate nesting

File: server/helpers/stakeholdersDirectory/stakeholderHelpers.js [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+        return stakeholder;
+      })
+      .catch((error) => {
+        throw error;
+      });
+
+    return { stakeholder };
+  } catch (e) {
+    if (!session.hasEnded) {
+      await session.endSession();
```
Comment (shemogumbe) [inline server/helpers/stakeholdersDirectory/stakeholderHelpers.js]: remove session and transactions

File: server/helpers/stakeholdersDirectory/static.js [hunk 5]
```diff
@@ -0,0 +1,281 @@
+import mongoose from '../../utils/secondaryMongooseConnection';
```
Comment (shemogumbe) [inline server/helpers/stakeholdersDirectory/static.js]: We would not need this if we are not using mongoose transactions

File: server/models/dropdowns/ThematicPillarDropdown.js [hunk 6]
```diff
@@ -1,14 +1,14 @@
...
-  pillarTitle: { type: String, unique: true },
-  description: { type: String, required: false },
-});
-
-ThematicPillarDropdown.defaultColumns = 'pillarTitle';
-
-ThematicPillarDropdown.register();
-
-export default ThematicPillarDropdown;
+// import keystone from 'keystone';
```
Comment (shemogumbe) [inline server/models/dropdowns/ThematicPillarDropdown.js]: All these sommented out code should be removed from source control

File: server/helpers/stakeholdersDirectory/stakeholderHelpers.js (hunk 1 above)
Comment (fgkinus) [reply server/helpers/stakeholdersDirectory/stakeholderHelpers.js]: This has been resolved and replaced by manual roll back

File: server/helpers/stakeholdersDirectory/stakeholderHelpers.js (hunk 2 above)
Comment (fgkinus) [reply server/helpers/stakeholdersDirectory/stakeholderHelpers.js]: I have done co cleanup and removed this debugging statements

**Comment by Ruiru11** [comment]
Really good work @fgkinus but there are a couple of things
1. I think if possible we should do validations on the form field itself like for the website url section you can only know its invalid when you are done and want to save 
image

2. The back button doesn't work as expected that means if I am on the second page and need to go to the first page it doesn't work like that

3. I could not understand this error
image

4. There are two impact-type sections
image

**Comment by fgkinus** [comment]
> Good job on this feature @fgkinus, I notice that there are two impact-type input field as you add stakeholder basic information.
> Screenshot 2019-04-30 at 10 18 35

Hi @Nicanor008 . That is what is present in the mockups. I will raise the issue with Chinedu and see what the way forward will be.
image

File: .elasticbeanstalk/config.yml [hunk 7]
```diff
@@ -0,0 +1,21 @@
+branch-defaults:
```
Comment (shemogumbe) [inline .elasticbeanstalk/config.yml]: Please remove this file as it is not needed anywhere, deployment uses travisCI, delete it, commit then add to gitignore and commit again

File: .travis.yml [hunk 8]
```diff
@@ -34,4 +34,4 @@ deploy:
     branch: develop
   access_key_id: $AWS_ACCESS_KEY
   secret_access_key:
-    secure: $AWS_SECRET_KEY
+secure: $AWS_SECRET_KEY
```
Comment (shemogumbe) [inline .travis.yml]: Avoid any update to TRAVISci file in this PR, revert this and restore to the one in develop

**Comment by fgkinus** [comment]
> 'Add stakeholder' container should be centred on the page
> Screenshot 2019-04-30 at 10 45 51

I have fixed this alignment to be more centred
image

**Pull request merged**